module/zfs/arc.c

```diff
@@ -7743,12 +7883,22 @@ arc_init(void)
 
 	buf_init();
 
+	if (zfs_arc_evict_threads == 0)
+		zfs_arc_evict_threads_live = MIN(MAX(max_ncpus > 6 ? 2 : 1,
+		    arc_ilog2(max_ncpus) + (max_ncpus >> 6)), 16);
```

Comment:
Do we really need the `MAX(max_ncpus > 6 ? 2 : 1` part? ilog2(6) should already be 2.

Reply:
Currently, we get the following thread counts:

| CPU count | Resulting thread count |
|---|---|
| 1 |  1 |
| 2 | 1 |
|3| 1|
|4 | 2|
| ... | ....|

So you are right that the `MAX(max_ncpus > 6 ? 2 : 1` part is not working as expected currently. If we want to stick to 1 thread for 4 and 5 CPUs then we'd need to use the following formula:

```c
// Version 2a
MIN(max_ncpus < 6 ? 1 : arc_ilog2(max_ncpus) + (max_ncpus >> 6), 16);
```

If we decide to simplify that further and go for 2 threads on systems with 4 or 5 CPUs, then we can just use:

```c
// Version 2b
MAX(1, MIN(arc_ilog2(max_ncpus) + (max_ncpus >> 6), 16));
```

Version 2b looks good to me and is certainly easier to reason about.

Reply:
2b looks cleaner, but 2 threads out of 4 sound a bit overkill to me.  May be the curve could be thought more.  BTW, speaking about more clean (readable code), this is not performance-critical part and we are not in 1990's, there is no point to use bit shifts for division, you may just use ` / 64` and compilers will do it right. And then you would not need parenthesis around it.

Reply:
Thank you for your feedback! I've cleaned up the formula. Now, we use 1 thread for less than 6 CPUs and then `MIN((highbit64(max_ncpus) - 1) + max_ncpus / 64, 16)`. for larger systems.

Reply:
Thanks.  Just it seems like you are still creating a taskq with one thread, but never using it. ;)

Reply:
Good catch. Thank you for pointing that out. 

I've fixed that. Now all evict-taskq-related bits are wrapped with a check if the use of the taskq is on.

Reply:
Thanks. That's fine, but I personally would check for `arc_evict_taskq` being NULL rather than !`live > 1`.